### PR TITLE
DOMA-6020 Call records attachment

### PR DESCRIPTION
**Pull request by nomerdvadcatpyat**
Some fixes for telephony integration

### CODE REVIEW

File: apps/condo/migrations/20230529194309-0275_callrecord_callrecordfragmenthistoryrecord_and_more.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+CREATE INDEX "CallRecordFragment_organization_66aac46b" ON "CallRecordFragment" ("organization");
+CREATE INDEX "CallRecordFragment_ticket_7c7d9ec7" ON "CallRecordFragment" ("ticket");
+CREATE INDEX "CallRecordFragment_updatedBy_e382c27e" ON "CallRecordFragment" ("updatedBy");
+CREATE INDEX "CallRecord_organization_0525fbb1" ON "CallRecord" ("organization");
+CREATE INDEX "CallRecord_updatedBy_741d8710" ON "CallRecord" ("updatedBy");
+
+UPDATE "OrganizationEmployeeRole"
+SET "canManageCallRecords" = true
+WHERE "name" = 'employee.role.Administrator.name'
+    OR "name" = 'employee.role.Dispatcher.name'
```
Comment (Alllex202) [inline apps/condo/migrations/20230529194309-0275_callrecord_callrecordfragmenthistoryrecord_and_more.js:86]: should mark as `custom`

File: apps/condo/domains/ticket/schema/CallRecord.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            isRequired: true,
+            knexOptions: { isNotNullable: true },
+            kmigratorOptions: { null: false, on_delete: 'models.CASCADE' },
+        },
+        file: {
+            schemaDoc: 'Conversation file',
+            type: File,
+            adapter: Adapter,
+        },
+        callerPhone: {
```
[the remark this answers is not in the document]
Comment (nomerdvadcatpyat) [reply apps/condo/domains/ticket/schema/CallRecord.js:37]: added

File: apps/condo/domains/ticket/schema/CallRecord.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            schemaDoc: 'Phone number of the person who called',
+            type: Text,
+            isRequired: true,
+        },
+        destCallerPhone: {
```
[the remark this answers is not in the document]
Comment (nomerdvadcatpyat) [reply apps/condo/domains/ticket/schema/CallRecord.js:54]: added

File: apps/condo/domains/ticket/access/CallRecordFragment.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+const { isSoftDelete } = require('@open-condo/keystone/access')
+const { throwAuthenticationError } = require('@open-condo/keystone/apolloErrorFormatter')
+const { getById } = require('@open-condo/keystone/schema')
+
+const { checkPermissionInUserOrganizationOrRelatedOrganization, queryOrganizationEmployeeFor, queryOrganizationEmployeeFromRelatedOrganizationFor } = require('@condo/domains/organization/utils/accessSchema')
+
+async function canReadCallRecordFragments ({ authentication: { item: user } }) {
+    if (!user) return throwAuthenticationError()
+    if (user.deletedAt) return false
+    if (user.isAdmin || user.isSupport) return {}
```
Comment (pahaz) [inline apps/condo/domains/ticket/access/CallRecordFragment.js:16]: probably we don't need to manage it by support

File: apps/condo/domains/ticket/access/CallRecordFragment.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            ],
+        },
+    }
+}
+
+async function canManageCallRecordFragments ({ authentication: { item: user }, originalInput, operation, itemId }) {
+    if (!user) return throwAuthenticationError()
+    if (user.deletedAt) return false
+    if (operation === 'update' && !isSoftDelete(originalInput)) return false
+    if (user.isAdmin || user.isSupport) return true
```
Comment (pahaz) [inline apps/condo/domains/ticket/access/CallRecordFragment.js:32]: same, no support cases

File: apps/condo/domains/ticket/constants/errors.js [hunk 6]
```diff
@@ -1,5 +1,25 @@
...
+    INVALID_CALLER_PHONE_NUMBER_FORMAT: {
+        code: 'BAD_USER_INPUT',
+        type: INVALID_PHONE_NUMBER_FORMAT,
+        message: INVALID_PHONE_NUMBER_FORMAT_MESSAGE,
+        variable: ['data', 'callerPhone'],
+    },
+    INVALID_DEST_CALLER_PHONE_NUMBER_FORMAT: {
+        code: 'BAD_USER_INPUT',
+        type: INVALID_PHONE_NUMBER_FORMAT,
+        message: INVALID_PHONE_NUMBER_FORMAT_MESSAGE,
```
Comment (Alllex202) [inline apps/condo/domains/ticket/constants/errors.js:16]: same

File: apps/condo/domains/ticket/schema/CallRecord.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+                resolveInput: async ({ resolvedData }) => {
+                    const newValue = normalizePhone(resolvedData['destCallerPhone'], true)
+                    return newValue || resolvedData['destCallerPhone']
+                },
+            },
+        },
+        talkTime: {
+            schemaDoc: 'Time of conversation between operator and client (in seconds)',
+            type: Integer,
+            isRequired: true,
```
Comment (Alllex202) [inline apps/condo/domains/ticket/schema/CallRecord.js:62]: must be greater than or equal to 0

File: apps/condo/domains/ticket/schema/CallRecord.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        },
+    },
+    hooks: {
+        validateInput: async ({ resolvedData }) => {
+            const newCallerPhone = normalizePhone(resolvedData['callerPhone'], true)
+            const newDestCallerPhone = normalizePhone(resolvedData['destCallerPhone'], true)
+
+            if (resolvedData['callerPhone'] && newCallerPhone !== resolvedData['callerPhone']) {
+                throw new GQLError(CALL_RECORD_ERRORS.INVALID_CALLER_PHONE_NUMBER_FORMAT)
```
Comment (Alllex202) [inline apps/condo/domains/ticket/schema/CallRecord.js:86]: you should pass context

File: apps/condo/domains/ticket/schema/CallRecord.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            }
+            if (resolvedData['destCallerPhone'] && newDestCallerPhone !== resolvedData['destCallerPhone']) {
+                throw new GQLError(CALL_RECORD_ERRORS.INVALID_DEST_CALLER_PHONE_NUMBER_FORMAT)
```
Comment (Alllex202) [inline apps/condo/domains/ticket/schema/CallRecord.js:89]: same

Comment (Alllex202) [reply apps/condo/domains/ticket/schema/CallRecord.js:89]: and in other places too

File: apps/condo/domains/ticket/schema/CallRecordFragment.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+const { GQLListSchema } = require('@open-condo/keystone/schema')
+
+const { addOrganizationFieldPlugin } = require('@condo/domains/organization/schema/plugins/addOrganizationFieldPlugin')
+const access = require('@condo/domains/ticket/access/CallRecordFragment')
+
+
+const CallRecordFragment = new GQLListSchema('CallRecordFragment', {
+    schemaDoc: 'M2M relationship between ticket and call record',
+    fields: {
+        ticket: {
```
Comment (Alllex202) [inline apps/condo/domains/ticket/schema/CallRecordFragment.js:17]: should add schemaDoc

File: apps/condo/domains/ticket/schema/CallRecordFragment.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+            type: Relationship,
+            ref: 'Ticket',
+            knexOptions: { isNotNullable: false },
+        },
+        callRecord: {
```
Comment (Alllex202) [inline apps/condo/domains/ticket/schema/CallRecordFragment.js:22]: same

File: apps/condo/domains/organization/gql.js [hunk 12]
```diff
@@ -10,10 +10,10 @@ const { generateGqlQueries } = require('@open-condo/codegen/generate.gql')
 
 const COMMON_FIELDS = 'id dv sender { dv fingerprint } v deletedAt newId createdBy { id name } updatedBy { id name } createdAt updatedAt'
 
-const ORGANIZATION_FIELDS = `{ country name description avatar { publicUrl } meta tin features statusTransitions defaultEmployeeRoleStatusTransitions importId importRemoteSystem ${COMMON_FIELDS} }`
+const ORGANIZATION_FIELDS = `{ country name description avatar { publicUrl } meta tin features statusTransitions defaultEmployeeRoleStatusTransitions importId importRemoteSystem phone phoneNumberPrefix ${COMMON_FIELDS} }`
```
Comment (pahaz) [inline apps/condo/domains/organization/gql.js:13]: do you want to change `phoneNUmberProfix`? it's no so clear without context

Comment (nomerdvadcatpyat) [reply apps/condo/domains/organization/gql.js:13]: Updated context

File: apps/condo/domains/common/schema/fields.js [hunk 13]
```diff
@@ -163,6 +163,21 @@ const UNIT_TYPE_FIELD = {
     defaultValue: FLAT_UNIT_TYPE,
 }
 
+const DEFAULT_PHONE_SCHEMA_DOC = 'Normalized phone in E.164 format without spaces'
+
+const getPhoneField = ({ fieldName, schemaDoc = DEFAULT_PHONE_SCHEMA_DOC }) => {
```
Comment (Alllex202) [inline apps/condo/domains/common/schema/fields.js:168]: maybe it's better to use object format?

Comment (Alllex202) [reply apps/condo/domains/common/schema/fields.js:168]: so you can easily expand the fields

Comment (nomerdvadcatpyat) [reply apps/condo/domains/common/schema/fields.js:168]: updated

File: apps/condo/domains/common/schema/fields.js [hunk 14]
```diff
@@ -163,6 +163,19 @@ const UNIT_TYPE_FIELD = {
     defaultValue: FLAT_UNIT_TYPE,
 }
 
+const DEFAULT_PHONE_SCHEMA_DOC = 'Normalized phone in E.164 format without spaces'
```
Comment (Alllex202) [inline apps/condo/domains/common/schema/fields.js:166]: Without this variable it will be better

**Review by Alllex202 — APPROVED** [review]

File: apps/condo/domains/common/schema/fields.js [hunk 15]
```diff
@@ -163,6 +163,19 @@ const UNIT_TYPE_FIELD = {
...
+
+const PHONE_FIELD = {
+    schemaDoc: DEFAULT_PHONE_SCHEMA_DOC,
```
Comment (pahaz) [inline apps/condo/domains/common/schema/fields.js:169]: It’s better to hardcode the doc here

File: apps/condo/domains/ticket/constants/errors.js [hunk 16]
```diff
@@ -1,5 +1,36 @@
 const UNKNOWN_GROUP_BY_FILTER = 'UNKNOWN_GROUP_BY_FILTER'
+const INVALID_PHONE_NUMBER_FORMAT = 'INVALID_PHONE_NUMBER_FORMAT'
+const NEGATIVE_TALK_TIME = 'NEGATIVE_TALK_TIME'
+
+const INVALID_PHONE_NUMBER_FORMAT_MESSAGE = 'Invalid phone number format'
```
Comment (pahaz) [inline apps/condo/domains/ticket/constants/errors.js:5]: It’s better to use INVALID_CALLER_PHONE_NUMBER_FORMAT.message and hardcode this value

**Comment by pahaz** [comment]
You need to check that you can’t connect ticket from one organization to call record from another. 

We need to check the consistency of organization field

**Comment by nomerdvadcatpyat** [comment]
> You need to check that you can’t connect ticket from one organization to call record from another.
> 
> We need to check the consistency of organization field

Added validation

File: apps/condo/domains/common/schema/fields.js [hunk 17]
```diff
@@ -1,14 +1,17 @@
...
 const { Json, SignedDecimal } = require('@open-condo/keystone/fields')
 
 const { ISO_CODES } = require('@condo/domains/common/constants/currencies')
-const { PHONE_WRONG_FORMAT_ERROR, JSON_UNKNOWN_VERSION_ERROR, REQUIRED_NO_VALUE_ERROR, JSON_EXPECT_OBJECT_ERROR } = require('@condo/domains/common/constants/errors')
+const { PHONE_WRONG_FORMAT_ERROR, JSON_UNKNOWN_VERSION_ERROR, REQUIRED_NO_VALUE_ERROR, JSON_EXPECT_OBJECT_ERROR, COMMON_ERRORS } = require('@condo/domains/common/constants/errors')
 const { normalizePhone } = require('@condo/domains/common/utils/phone')
 const { UNIT_TYPES, FLAT_UNIT_TYPE } = require('@condo/domains/property/constants/common')
 const { ADDRESS_META_FIELD_GRAPHQL_TYPES, ADDRESS_META_SUBFIELDS_QUERY_LIST } = require('@condo/domains/property/schema/fields/AddressMetaField')
 
+const { CALL_RECORD_ERRORS } = require('../../ticket/constants/errors')
```
Comment (pahaz) [inline apps/condo/domains/common/schema/fields.js:13]: You can’t import ticket from common domain

File: apps/condo/domains/ticket/constants/errors.js [hunk 18]
```diff
@@ -1,5 +1,30 @@
 const UNKNOWN_GROUP_BY_FILTER = 'UNKNOWN_GROUP_BY_FILTER'
+const NEGATIVE_TALK_TIME = 'NEGATIVE_TALK_TIME'
```
Comment (pahaz) [inline apps/condo/domains/ticket/constants/errors.js:2]: It’s better to online

File: apps/condo/domains/common/schema/fields.js (hunk 17 above)
Comment (nomerdvadcatpyat) [reply apps/condo/domains/common/schema/fields.js:13]: It's unused import, removed it

**Pull request merged**